### PR TITLE
Add task instance mutation hook

### PR DESCRIPTION
**Pull request by milton0825**
tl;dr: support task instance mutation before each execution.

Currently the only way to apply cluster-wide mutations to tasks is by defining `policy` in `airflow_local_setting`. However, `policy` is applied to `task` (not `task_instance` and the descriptions and docs are wrong) right after the DAG is loaded. And you will not be able to change the task after that.

Therefore a hook is added to support mutating the task instance before each execution (in case there is a retry). For example, you can re-route the task to execute in a different queue during retries.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by feng-tao** [comment]
@ashb could you take a look? Lyft leverages this features to do py2/py3 migration and found it pretty useful. I haven't followed actively on recent airflow development and not sure if there is anything similar exist in current codebase /feature. Let us know.

File: airflow/models/dagrun.py [hunk 1]
```diff
@@ -431,6 +432,7 @@ def verify_integrity(self, session=None):
         # check for removed or restored tasks
         task_ids = []
         for ti in tis:
+            task_instance_mutation_hook(ti)
```
Comment (kaxil) [inline airflow/models/dagrun.py:436]: Why do we need it here? this TIs are received from DB.

I think we should only mutate before storing it to DB

Comment (milton0825) [reply airflow/models/dagrun.py:436]: For cases that we want to mutate task instances before each executions. For example, re-routing to different queue for retries.

Comment (kaxil) [reply airflow/models/dagrun.py:436]: But you are already adding it before saving it to DB on L468, so is it needed here too is my question?

Comment (milton0825) [reply airflow/models/dagrun.py:436]: L468 is reached when the task instance does not exist thus you create the task instance. However, during retries, the task instance already exists so it won't reach L468

Comment (kaxil) [reply airflow/models/dagrun.py:436]: aah yes if you want to have a different queue for TI's that are up for retry this does makes sense

File: tests/models/test_dagrun.py [hunk 2]
```diff
@@ -564,6 +565,27 @@ def with_all_tasks_removed(dag):
         flaky_ti.refresh_from_db()
         self.assertEqual(State.NONE, flaky_ti.state)
 
+    @mock.patch('airflow.models.dagrun.task_instance_mutation_hook')
```
Comment (kaxil) [inline tests/models/test_dagrun.py:569]: Can you add a test for TI where it is UP_FOR_RETRY

**Review by feng-tao — APPROVED** [review]

File: airflow/settings.py [hunk 3]
```diff
@@ -79,18 +84,13 @@
...
-
-    Note that the ``Task`` object has a reference to the DAG
-    object. So you can use the attributes of all of these to define your
-    policy.
+    This policy setting allows altering tasks after they are loaded in
+    the DagBag. It allows administrator to rewire some task parameters.
 
     To define policy, add a ``airflow_local_settings`` module
-    to your PYTHONPATH that defines this ``policy`` function. It receives
-    a ``Task`` object and can alter it where needed.
```
Comment (ashb) [inline airflow/settings.py:93]: Did you mean to remove these doc  comments? This bit seems appropriate still.

Comment (milton0825) [reply airflow/settings.py:93]: I moved that to the sections below.

File: tests/models/test_dagrun.py (hunk 2 above)
Comment (kaxil) [reply tests/models/test_dagrun.py:569]: @milton0825 Can you add a test for TI where it is UP_FOR_RETRY

Comment (milton0825) [reply tests/models/test_dagrun.py:569]: Testing that probably won't add too much value as the logic is not based on the state of TI.

Comment (kaxil) [reply tests/models/test_dagrun.py:569]: https://github.com/apache/airflow/pull/8852#discussion_r428262018 Based on our discussion in that comment if in future someone removes L468, no test would fail but it would break the behaviour that ti_mutation_hook won't work on ti that are up for retry.

Comment (milton0825) [reply tests/models/test_dagrun.py:569]: added
